scripts/aws_replicate.py

```diff
@@ -0,0 +1,376 @@
+import os
+import subprocess
+import shlex
+import Queue
+import time
+import timeit
+import threading
+
+import json
+import boto3
+
+from cdislogging import get_logger
+from indexclient.client import IndexClient
+
+from errors import APIError
+from settings import PROJECT_MAP, INDEXD
+from utils import (
+    get_bucket_name,
+    get_fileinfo_list_from_manifest,
+    get_fileinfo_list_from_s3_manifest,
+    exec_files_grouping,
+)
+
+logger = get_logger("AWSReplication")
+
+global EXIT_FLAG
+EXIT_FLAG = 0
+
+global TOTAL_PROCESSED_FILES
+TOTAL_PROCESSED_FILES = 0
+
+mutexLock = threading.Lock()
+workQueue = Queue.Queue()
+
+indexclient = IndexClient(
+    INDEXD["host"],
+    INDEXD["version"],
+    (INDEXD["auth"]["username"], INDEXD["auth"]["password"]),
+)
+
+
+class singleThread(threading.Thread):
+    def __init__(
+        self,
+        threadID,
+        threadName,
+        source_bucket,
+        copied_objects,
+        global_config,
+        job_name,
+        q,
+    ):
+        threading.Thread.__init__(self)
+        self.threadID = threadID
+        self.threadName = threadName
+        self.global_config = global_config
+        self.q = q
+        self.source_bucket = source_bucket
+        self.copied_objects = copied_objects
+        self.job_name = job_name
+        self.json_log = {}
+
+    def run(self):
+        logger.info("Starting " + self.name)
+        if self.job_name == "copying":
+            process_data(
+                self.threadName,
+                self.source_bucket,
+                self.global_config,
+                self.copied_objects,
+                self.q,
+            )
+        elif self.job_name == "indexing":
+            self.indexing_data(
+                self.threadName, self.global_config, self.copied_objects, self.q
+            )
+        logger.info("\nExiting " + self.name)
+
+    def indexing_data(self, threadName, global_config, copied_objects, q):
+        """
+        Start to index the copied object
+
+        Args:
+            threadName(str): thread name
+            global_config(dict): configuration dictionary
+            copied_objects(set): objects aready copied
+            q(Queue): queue containing the objects need to be processed
+        
+        Return:
+            None
+        """
+
+        while not EXIT_FLAG:
```

Comment:
I think this logic with the globals is messy for a number of reasons. `EXIT_FLAG` should be replaced with `thread.exit()` for sure, and you could change `run` to return the number of processed objects and aggregate these in the replicator's `run`